File: Y3/Project Management System/Java class files/Project.java
import org.jetbrains.annotations.NotNull;
import java.util.ArrayList;
import java.time.LocalDate;

public class Project {
    private String projName;//Name of the project
    private String projNote;//Brief description of the project

    private ArrayList<Task> ProjectTasks = new ArrayList<>();//Stores the tasks required for project completion
    private LocalDate StartDate;//date project was created
    private int ProjDuration;//deadline date that the project finishes
    private Task CurrentTask;

    //constructor
    public Project(String name, String note, String startdate, int dur){
        projName = name;
        projNote = note;
        StartDate = LocalDate.parse(startdate);
        ProjDuration = dur;
    }

    public void addTask(String desc, int d){
        Task tmp = new Task(desc, Project.this, d);
        this.ProjectTasks.add(tmp);
        tmp.checkPosition();
        for (int i =0; i < this.ProjectTasks.size()-1; i++){
            Task t = ProjectTasks.get(i);
            t.checkPosition();
        }
    }

    public void findCurrentTask(){
        for(int i = 0; i < this.ProjectTasks.size(); i++){
            Boolean check = this.ProjectTasks.get(i).getIsCompleted();
            if (check == false){
                this.CurrentTask = this.ProjectTasks.get(i);
                break;
            }
        }
    }


    public ArrayList<Task> getProjectTasks(){
        return ProjectTasks;
    }

    public void displayTaskData(){
        for (int i =0; i < ProjectTasks.size(); i++){
            Task tmp = ProjectTasks.get(i);
            System.out.println("Description: " + tmp +" "+ tmp.getTaskDesc() + " Start Date: " + tmp.getStartDate());
        }
    }

    public String getProjName(){
        return projName;
    }//returns project name

    public void setProjName(@NotNull String a){//sets the project name
        if (a.length() > 20){
            System.out.print("Error: project name exceeds length");
        } else{
            projName = a;
        }
    }

    public LocalDate getStartDate(){
        return StartDate;
    }


    public Task getCurrentTask() {
        return CurrentTask;
    }
}

File: Y3/Project Management System/src/classes/Project.kt
package classes

import java.io.Serializable
import java.time.LocalDate
import java.util.*

class Project(private var projName: String, val projNote: String, startdate: String?, dur: Int) : Serializable {
    val projectTasks = ArrayList<Task>()
    val projectTeams = ArrayList<Team>()
    val startDate: LocalDate
    val projDuration: Int
    var currentTask: Task? = null
        private set
    @JvmField
    var nonAssigned = Team("N/A")
    fun addTask(desc: String, d: Int, addTo: String) {
        val tmp = Task(desc, this, d, addTo)
        projectTasks.add(tmp)
        tmp.checkPosition()
        for (i in 0 until projectTasks.size - 1) {
            val t = projectTasks[i]
            t.checkPosition()
        }
        findCurrentTask()
    }

    fun findCurrentTask() {
        for (task in projectTasks) {
            val check = task.isCompleted
            if (!check) {
                currentTask = task
                break
            }
        }
    }

    fun createTeam(name: String, desc: String, n: Int) {
        val tmp = Team(name, desc, n)
        projectTeams.add(tmp)
    }

    fun AssignTeamToTask(team: String, task: String) {
        val te = getTeamOfName(team)
        val tk = getTaskOfDescription(task)
        if (te == null || tk == null) {
            println("ERROR: cannot find team or task")
        } else {
            tk.assignedTeam = te
            te.assignTask(tk)
        }
    }

    fun getTeamOfName(name: String): Team? {
        var a: Team? = null
        for (projectTeam in projectTeams) {
            if (projectTeam.getTeamName() == name) {
                a = projectTeam
                break
            }
        }
        return a
    }

    fun getTaskOfDescription(description: String): Task? {
        var a: Task? = null
        for (projectTask in projectTasks) {
            if (projectTask.taskDesc == description) {
                a = projectTask
                break
            }
        }
        return a
    }

    fun getProjName(): String {
        return projName
    } //returns project name

    fun setProjName(a: String) { //sets the project name
        if (a.length > 20) {
            print("Error: project name exceeds length")
        } else {
            projName = a
        }
    }

    //constructor
    init {
        startDate = LocalDate.parse(startdate)
        projDuration = dur
        val expdEndDate = startDate.plusDays(projDuration.toLong())
    }
}